File: client/lib/display_track.py
"""Python port of viewer/src/lib/displayTrack.ts.

Builds the same dwell + similarity + preempt schedule the viewer uses.
The TTS renderer consumes this so it speaks against the same blocks the
viewer flashes amber on. Drift between this file and displayTrack.ts is a
bug; tests in test_display_track.py mirror the TS test suite.
"""

import math
import re
from dataclasses import dataclass
from typing import Any

from .urgency import (
    classify_advice,
    contains_urgency_keyword,
    escalates,
    split_observation_advice,
)

DEFAULT_MIN_DWELL_MS = 2500
DEFAULT_JACCARD_THRESHOLD = 0.5

_TOKEN_PUNCT_RE = re.compile(r"[^\w\s]", flags=re.UNICODE)


@dataclass
class DisplayBlock:
    start_time: float
    end_time: float  # math.inf for the last block
    text: str
    source_entry: dict[str, Any]
    was_preempted: bool


def _tokenise(observation: str) -> set[str]:
    cleaned = _TOKEN_PUNCT_RE.sub(" ", observation.lower())
    return {w for w in cleaned.split() if len(w) > 2}


def _jaccard(a: set[str], b: set[str]) -> float:
    if not a and not b:
        return 1.0
    intersection = len(a & b)
    union = len(a) + len(b) - intersection
    return 1.0 if union == 0 else intersection / union


def _meaningfully_different(
    prev: dict[str, Any], nxt: dict[str, Any], jaccard_threshold: float
) -> bool:
    prev_obs, prev_adv = split_observation_advice(prev["text"])
    nxt_obs, nxt_adv = split_observation_advice(nxt["text"])
    if prev_adv != nxt_adv:
        return True
    return _jaccard(_tokenise(prev_obs), _tokenise(nxt_obs)) < jaccard_threshold


def _should_preempt(current: dict[str, Any], candidate: dict[str, Any]) -> bool:
    current_class = classify_advice(current["text"])
    candidate_class = classify_advice(candidate["text"])
    if escalates(current_class, candidate_class):
        return True
    # Observation-keyword preempt fires only when the candidate has a keyword
    # the current displayed caption did not — sustained collisions don't cause
    # repeated preempts.
    return contains_urgency_keyword(candidate["text"]) and not contains_urgency_keyword(
        current["text"]
    )


def build_display_track(
    entries: list[dict[str, Any]],
    *,
    min_dwell_ms: int = DEFAULT_MIN_DWELL_MS,
    jaccard_threshold: float = DEFAULT_JACCARD_THRESHOLD,
) -> list[DisplayBlock]:
    """Build a display schedule from raw session entries.

    Assumes `entries` is sorted by `capture_t_s` ascending. Output blocks have
    non-overlapping [start_time, end_time) intervals; the last block has
    end_time = math.inf.
    """
    track: list[DisplayBlock] = []
    if not entries:
        return track

    min_dwell_s = min_dwell_ms / 1000.0

    def open_block(entry: dict[str, Any], was_preempted: bool) -> None:
        if track:
            track[-1].end_time = entry["capture_t_s"]
        track.append(
            DisplayBlock(
                start_time=entry["capture_t_s"],
                end_time=math.inf,
                text=entry["text"],
                source_entry=entry,
                was_preempted=was_preempted,
            )
        )

    for entry in entries:
        if not track:
            open_block(entry, was_preempted=False)
            continue
        current = track[-1]
        elapsed = entry["capture_t_s"] - current.start_time

        if elapsed < min_dwell_s:
            # Inside dwell. The only way to swap is a hard-preempt.
            if _should_preempt(current.source_entry, entry):
                open_block(entry, was_preempted=True)
            continue

        # Past dwell.
        if _meaningfully_different(current.source_entry, entry, jaccard_threshold):
            open_block(entry, was_preempted=False)
        # else: extend current block (no-op; end_time stays inf).

    return track

File: viewer/src/lib/displayTrack.ts
import {
  DEFAULT_DISPLAY_CONFIG,
  type DisplayConfig,
  type DisplayEntry,
  type SessionEntry,
} from "../types";
import {
  classifyAdvice,
  containsUrgencyKeyword,
  escalates,
  splitObservationAdvice,
} from "./urgency";

// Tokenise an observation into a Set of lowercase word stems for Jaccard
// comparison. Punctuation is stripped; very short tokens are dropped.
function tokenise(observation: string): Set<string> {
  return new Set(
    observation
      .toLowerCase()
      .replace(/[^\p{L}\p{N}\s]/gu, " ")
      .split(/\s+/)
      .filter((w) => w.length > 2),
  );
}

function jaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 1;
  let intersection = 0;
  for (const x of a) if (b.has(x)) intersection++;
  const union = a.size + b.size - intersection;
  return union === 0 ? 1 : intersection / union;
}

function meaningfullyDifferent(
  prev: SessionEntry,
  next: SessionEntry,
  jaccardThreshold: number,
): boolean {
  const prevSplit = splitObservationAdvice(prev.text);
  const nextSplit = splitObservationAdvice(next.text);
  if (prevSplit.advice !== nextSplit.advice) return true;
  return (
    jaccard(tokenise(prevSplit.observation), tokenise(nextSplit.observation)) <
    jaccardThreshold
  );
}

function shouldPreempt(
  current: SessionEntry,
  candidate: SessionEntry,
): boolean {
  const currentClass = classifyAdvice(current.text);
  const candidateClass = classifyAdvice(candidate.text);
  if (escalates(currentClass, candidateClass)) return true;
  // Observation-keyword preempt fires only when the candidate has a keyword
  // the current displayed caption did not — sustained collisions don't cause
  // repeated preempts.
  return (
    containsUrgencyKeyword(candidate.text) &&
    !containsUrgencyKeyword(current.text)
  );
}

/**
 * Build a display schedule from raw session entries.
 *
 * Assumes `entries` is sorted by `capture_t_s` ascending (the recorder writes
 * them in this order).
 *
 * The output is a list of `DisplayEntry` blocks with non-overlapping
 * [startTime, endTime) intervals; the last block has `endTime: Infinity`.
 *
 * See plan: "Subtitle selection — display track with urgency override".
 */
export function buildDisplayTrack(
  entries: SessionEntry[],
  config: DisplayConfig = DEFAULT_DISPLAY_CONFIG,
): DisplayEntry[] {
  const track: DisplayEntry[] = [];
  if (entries.length === 0) return track;

  const minDwellS = config.minDwellMs / 1000;

  // Helper: open a new block at the given entry. When closing the previous
  // block (if any), set endTime to entry.capture_t_s.
  const open = (entry: SessionEntry, wasPreempted: boolean) => {
    if (track.length > 0) {
      track[track.length - 1].endTime = entry.capture_t_s;
    }
    track.push({
      startTime: entry.capture_t_s,
      endTime: Infinity,
      text: entry.text,
      sourceEntry: entry,
      wasPreempted,
    });
  };

  for (const entry of entries) {
    if (track.length === 0) {
      open(entry, false);
      continue;
    }
    const current = track[track.length - 1];
    const elapsed = entry.capture_t_s - current.startTime;

    if (elapsed < minDwellS) {
      // Inside dwell. The only way to swap is a hard-preempt.
      if (shouldPreempt(current.sourceEntry, entry)) {
        open(entry, true);
      }
      // Otherwise drop the entry.
      continue;
    }

    // Past dwell.
    if (
      meaningfullyDifferent(current.sourceEntry, entry, config.jaccardThreshold)
    ) {
      open(entry, false);
    }
    // else: extend current block (no-op; endTime stays Infinity).
  }

  return track;
}

// Find the active display block at video time t.
// Returns null when t is before the first block's startTime.
export function activeDisplayAt(
  track: DisplayEntry[],
  t: number,
): DisplayEntry | null {
  if (track.length === 0) return null;
  if (t < track[0].startTime) return null;
  // Binary search for the rightmost block with startTime <= t.
  let lo = 0;
  let hi = track.length - 1;
  while (lo < hi) {
    const mid = (lo + hi + 1) >>> 1;
    if (track[mid].startTime <= t) lo = mid;
    else hi = mid - 1;
  }
  const block = track[lo];
  return t < block.endTime ? block : null;
}
